src/pages/settings/tabs/GeneralSettings.tsx

```diff
@@ -27,8 +32,49 @@ function GeneralSettings() {
     { channel: newSettings.WebUI_Settings.updateChannel, force: true },
     false,
   );
+  const themePathHref = useMemo(() => document.getElementById('theme-css')!.attributes.getNamedItem('href')!, []);
+  const fileInputRef = useRef<HTMLInputElement | null>(null);
   const versionQuery = useVersionQuery();
   const themesQuery = useWebuiThemesQuery();
+  const { isPending: isUploading, mutate: uploadTheme } = useWebuiUploadThemeMutation();
+
+  const onOpenFileDialog = useEventCallback((event: React.SyntheticEvent) => {
+    if (isUploading) {
+      return;
+    }
+
+    event.preventDefault();
+    event.stopPropagation();
+    if (fileInputRef.current) {
+      fileInputRef.current.click();
+    }
+  });
+
+  const onFileChange = useEventCallback((event: React.ChangeEvent<HTMLInputElement>) => {
+    if (isUploading) {
+      return;
+    }
+
+    event.preventDefault();
+    event.stopPropagation();
+    const file = event.target.files?.[0];
+    if (!file) return;
+    uploadTheme({ file }, {
+      async onSuccess(data) {
+        await themesQuery.refetch();
+
+        let path = themePathHref.value;
+        const index = path.indexOf('?');
+        if (index !== -1) path = path.substring(0, index);
+        themeUpdateCounter += 1;
+        path += `?updateCount=${themeUpdateCounter}`;
+        themePathHref.value = path;
```

Comment:
do this with `new URL(themePathHref).value` and corresponding properties, so that it's more readable